python/astshim/fitsChanContinued.py

```diff
@@ -0,0 +1,194 @@
+            return True
+
+    return False
+
+
+FitsChan.__contains__ = contains
+
+
+def getitem(self, name):
+    """Return the value if a keyword is specified, or the entire card if
+    an integer index is specified."""
+
+    # Save current card position
+    currentCard = self.getCard()
+
+    if isinstance(name, int):
+        # FITS is 1-based and Python is zero-based so add 1
+        self.setCard(name + 1)
+        result = self.findFits("%f", False)
+        self.setCard(currentCard)
+        if not result.found:
+            raise IndexError(f"No FITS card at index {name}")
+        return result.value
+
+    elif isinstance(name, str):
+        # Rewind FitsChan so we search all cards
+        self.clearCard()
+
+        # Method look up table for obtaining values
+        typeLut = {CardType.INT: self.getFitsI,
+                   CardType.FLOAT: self.getFitsF,
+                   CardType.STRING: self.getFitsS,
+                   CardType.COMPLEXF: self.getFitsCF,
+                   CardType.LOGICAL: self.getFitsL
+                   }
+
+        # We can have multiple matches
+        values = []
+
+        # Loop over each item that matches
+        while True:
+            result = self.findFits(name, False)
+            if not result.found:
+                break
+
+            # Get the data type for this matching card
+            ctype = self.getCardType()
+
+            # Go back one card so we can ask for the value in the correct
+            # data type (getFitsX starts from the next card)
+            thiscard = self.getCard()
+            self.setCard(thiscard - 1)
+
+            if ctype == CardType.UNDEF:
+                values.append(None)
+            elif ctype in typeLut:
+                found = typeLut[ctype](name)
+                if found.found:
+                    values.append(found.value)
+                else:
+                    raise RuntimeError(f"Unexpectedly failed to find card '{name}'")
+            else:
+                raise RuntimeError(f"Type, {ctype} of FITS card '{name}' not supported")
+
+            # Increment the card number to continue search
+            self.setCard(thiscard + 1)
+
+        # Reinstate the original card position
+        self.setCard(currentCard)
+
+        # We may have multiple values. For consistency with pyast we return
+        # a single item for a single match, else a tuple
+        result = None
+        if not values:
+            raise KeyError(f"{name}")
+
+        if len(values) == 1:
+            result = values[0]
+        else:
+            result = tuple(values)
+
+        return result
+
+    raise ValueError(f"Supplied key, '{name}' of unsupported type")
+
+
+FitsChan.__getitem__ = getitem
+
+
+def setitem(self, name, value):
+    """name can be integer index or keyword.
+
+    If an integer, the value is deemed to be a full card to replace the
+    existing one. If the value is None or empty string the card is deleted.
```

Comment:
The distinction is that if you set a card by position (ie with an integer index) to `None` it is deleted. If you set it by name to `None` you get an undefined value. Assigning `None` to a position is generally ambiguous since there is no undef equivalent (remember that assigning by index means you are assigning an entire card, not a value for the header item). A completely blank header would be an empty string.  I think the current logic is reasonable in this case.

Reply:
The doc string says "If the vlaue is None or empty string the card is deleted". And you just said in your comment "If you set it by name to `None` you get an undefined value`. The latter (keeping the card but making the value undefined) seems like much more desirable behavior.

Reply:
The first paragraph was solely talking about the integer index case. The next paragraph talks about the header name case. Regardless, I have removed the delete and added `__delitem__`.